src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1388 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <credentials/GroupDataProviderImpl.h>
+#include <credentials/PersistentStorageKey.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/Pool.h>
+#include <stdlib.h>
+#include <string.h>
+
+namespace chip {
+namespace Credentials {
+
+static constexpr uint16_t kPersistentBufferMax = 128;
+
+template <size_t S>
+struct PersistentData
+{
+public:
+    virtual ~PersistentData() = default;
+
+    virtual void UpdateKey()                                    = 0;
+    virtual void Clear()                                        = 0;
+    virtual CHIP_ERROR Serialize(TLV::TLVWriter & writer) const = 0;
+    virtual CHIP_ERROR Deserialize(TLV::TLVReader & reader)     = 0;
+
+    CHIP_ERROR Save(chip::PersistentStorageDelegate & storage)
+    {
+        // Serialize the data
+        TLV::TLVWriter writer;
+        writer.Init(mBuffer, sizeof(mBuffer));
+        ReturnErrorOnFailure(Serialize(writer));
+
+        // Update storage key
+        UpdateKey();
+        // Save serialized data
+        ReturnErrorOnFailure(storage.SyncSetKeyValue(mKey.Value(), mBuffer, static_cast<uint16_t>(writer.GetLengthWritten())));
+
+        // WARNING: This read-back is a walkaround for a bug in the Darwing implementation of KeyValueStoreMgr
+        //          See https://github.com/project-chip/connectedhomeip/issues/12174.
+        uint16_t size = static_cast<uint16_t>(sizeof(mBuffer));
+        ReturnErrorOnFailure(storage.SyncGetKeyValue(mKey.Value(), mBuffer, size));
+
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR Load(chip::PersistentStorageDelegate & storage)
+    {
+        // Update storage key
+        UpdateKey();
+        // Load the serialized data
+        uint16_t size = static_cast<uint16_t>(sizeof(mBuffer));
+        ReturnErrorOnFailure(storage.SyncGetKeyValue(mKey.Value(), mBuffer, size));
+
+        // Decode serialized data
+        Clear();
+        TLV::TLVReader reader;
+        reader.Init(mBuffer, size);
+        ReturnErrorOnFailure(Deserialize(reader));
+
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR Delete(chip::PersistentStorageDelegate & storage)
+    {
+        // Update storage key
+        UpdateKey();
+        // Delete stored data
+        return storage.SyncDeleteKeyValue(mKey.Value());
+    }
+
+protected:
+    PersistentStorageKey mKey;
```

Comment:
Suggest not owning this, otherwise each class that has a PersistentStorageKey instance will cost 32+1 bytes + padding bytes of RAM forever to hold the temp key buffer.

There may be a way to re-organize the derived classes to call a parent's method to actually store/load. In other words: `Store --> Serialize --> ActuallyStore(key, buf)` where ActuallyStore is in the parent class, The key name is ephemeral and never needs to stick-around. In a sense, mBuffer as well...

Reply:
Actually, it looks like there are no instances of `PersistentData` ever used outside of a method's stack. This may be fine for now, although if you have 2-3 `PeristentData` concrete types around in a method (like in `RemoveGroupMapping`), the stack cost will be large for the whole duration of the method.

Reply:
Now both key and buffer are allocated in the stack of the Load and Save functions.